### ISSUE TITLE
warning: value assigned to `x` is never read not fixed by cargo fix

### ISSUE THREAD

**Issue by jrmuizel** [rust-lang/rust#76847]
I tried this code:

```rust
fn main() {
    let mut x = 0;
    x = 8;
    println!("Hello, world! {}", x);
}
```

which gives:
```
warning: value assigned to `x` is never read
 --> src/main.rs:2:13
  |
2 |     let mut x = 0;
  |             ^
  |
  = note: `#[warn(unused_assignments)]` on by default
  = help: maybe it is overwritten before being read?
```

It would be nice if this warning was fixable with `cargo fix`

**Comment by newca12** [comment]
https://doc.rust-lang.org/cargo/commands/cargo-fix.html
If you encounter any problems with cargo fix or otherwise have any questions or feature requests please don’t hesitate to file an issue at https://github.com/rust-lang/cargo

**Comment by jrmuizel** [comment]
I filed this here because I don't think the problem is in `cargo fix`. I think `rustc` needs to suggest a fix and then `cargo fix` will just work.

**Comment by jyn514** [comment]
Relevant code: https://github.com/rust-lang/rust/blob/3e14b684dd31794e8dbe6acd7a5f132997ee0c47/compiler/rustc_passes/src/liveness.rs#L1774-L1793
Seems like the hard part will be coming up with a suggestion, maybe just to remove the assignment?

**Comment by lcnr** [comment]
We can't just remove the assignment here, see
```rust
fn weird<T>(v: T) -> T {
    println!("hey");
    v
}

fn main() {
    let mut x = weird(0);
    x = 8;
}
```
We could maybe recommend
```
let mut x; wierd(0); 
```
here, which should result in followup warnings, not sure if that's too great though

Or we only add the suggestion if the input is known to not have sideeffects, which is probably better :thinking:
